[atoms] support for inline styles

mellyeliu opened the pull request:
## Implementation

Adding support for inline styles authoring with a new `@stylexjs/atoms` package. This allows you to write styles inline in `stylex.props()` that are pre-compiled within the props visitor using much of the `create` logic for static and dynamic styles.

This is something we went back and forth on a few times, but ultimately felt that it's good to provide the optionality. We still recommend (and want to bake in enough friction) for people to use `stylex.create` namespaces for the majority of usecases for readability/maintainability reasons. 


```tsx
import * as stylex from '@stylexjs/stylex';
import x from '@stylexjs/atoms';

<div {...stylex.props(
  x.display.flex,
  x.padding._16px,
  x.width['calc(100% - 20cqi)'],
  x.color(color),
)} />
```

The naming of all the above is up for discussion, as well as what styles we should build into this API. I'm also using `Properties` type from estree for type checking for a first pass, but there might be something better we can use.

## Testing

Style types:
- [x] Static values 
- [x] Leading underscores 
- [x] Object key syntax
- [x] Dynamic inline styles


Tested within docs page the above +
- [x] Different merges of dynamic, regular, and inline styles
- [x] Named imports 

Deferring to follow ups:
- [ ] Rethink contextual styles
- [ ] Make work with valid-styles / some way to add linter validation
- [ ] Add docs
- [ ] Rethink package name + standard import alias
- [ ] figure out import and package naming?

nmn commented on packages/@stylexjs/inline-css/src/index.js:
```diff
@@ -0,0 +1,35 @@
+    apply() {
+      return undefined;
+    },
+  });
```
It should probably not be a recursive proxy and the `get()` should throw an error similar to the stylex package saying you need to compile this away.

nmn commented on packages/@stylexjs/inline-css/src/index.js:
```diff
@@ -0,0 +1,35 @@
+    if (typeof prop === 'string') {
+      return valueProxy(prop);
+    }
+    return undefined;
```
You can probably just throw here directly and not define `valueProxy` at all.

nmn commented on packages/@stylexjs/babel-plugin/__tests__/transform-stylex-props-test.js:
```diff
@@ -231,6 +231,451 @@ describe('@stylexjs/babel-plugin', () => {
+        const inline = transform(`
+          import stylex from 'stylex';
+          import * as css from '@stylexjs/inline-css';
+          stylex.props(css.display.flex);
```
```suggestion
          import {css} from '@stylexjs/inline-css';
          stylex.props(css.display.flex);
```

nmn commented on packages/@stylexjs/babel-plugin/__tests__/transform-stylex-props-test.js:
```diff
@@ -231,6 +231,451 @@ describe('@stylexjs/babel-plugin', () => {
+          stylex.props([{
+            "color": color != null ? "x14rh7hd" : color,
+            "$$css": true
+          }, {
```
It is important that that the style objects created by the inline-css package are hoisted to the module level as a constant and not inlined like this. We already have a utility for doing this somewhere.

nmn reviewed (dismissed):
Good start, but a few changes are needed.

nmn commented on packages/@stylexjs/babel-plugin/src/utils/state-manager.js:
```diff
@@ -165,12 +166,20 @@ export default class StateManager {
   // `stylex.create` calls
   +styleMap: Map<string, CompiledNamespaces> = new Map();
   +styleVars: Map<string, NodePath<>> = new Map();
+  +inlineStylesCache: Map<string, FlatCompiledStyles> = new Map();
```
What is this? Why is it needed?

nmn commented on packages/@stylexjs/babel-plugin/src/visitors/imports.js:
```diff
@@ -12,6 +12,8 @@ import type { NodePath } from '@babel/traverse';
 import * as t from '@babel/types';
 import StateManager from '../utils/state-manager';
 
+const UTILITY_STYLES_SOURCES = new Set(['@stylexjs/utility-styles']);
```
`/style-properties`?

vincentriemer commented on packages/@stylexjs/atoms/src/babel-transform.js:
```diff
@@ -0,0 +1,249 @@
+function isUtilityStylesIdentifier(ident, state, path) {
+  if (state.atomImports && state.atomImports.has(ident.name)) {
+    return true;
+  }
```
I think if the import's name is shadowed (i.e. you import the name `css`, but you're in a function that has an argument named `css`) it would still cause the replacement to happen.

mellyeliu committed:
[atoms] Add @stylexjs/atoms package for inline atomic styles

Add a new @stylexjs/atoms package that provides compile-time atomic style
helpers. Atoms compile directly to {$$css: true, ...} objects, so they
work everywhere: inline, as variables, through props, in arrays.

Usage:
  import x from '@stylexjs/atoms';
  x.display.flex        // static
  x.padding._16px       // leading underscore for numeric-start values
  x.width['calc(100%)'] // bracket syntax for complex values
  x.color(variable)     // dynamic via CSS custom properties

Babel plugin integration:
- Atoms visitor runs before stylex.props, compiling atoms in place
- Compilation utilities are injected from babel-plugin into atoms transform
- Atom imports tracked via state.atomImports in StateManager
- Re-exported as `x` from @stylexjs/stylex for convenience

mellyeliu committed:
[atoms] Add babel-plugin tests for atoms compilation

mellyeliu committed:
[atoms] Fix $ReadOnly→Readonly in Flow types, remove x re-export

- Use modern Flow Readonly<> syntax in index.js.flow
- Remove `export { default as x } from '@stylexjs/atoms'` from stylex.js
  to avoid circular dependency (atoms peer-depends on stylex)
- Users import atoms directly: `import x from '@stylexjs/atoms'`

mellyeliu committed:
Fix scripts devDependency version to match workspace (0.18.3)

mellyeliu committed:
[atoms] Add bailout and hoisting tests for atoms

- atoms mixed with unknown external styles (forces runtime bailout)
- atoms with dynamic conditional inside function (hoisted function pattern)
- atoms with variable reference (xstyle prop pattern, forces bailout)
